File: model-database/src/main/java/com/bnorm/barkeep/db/BookEntity.java
package com.bnorm.barkeep.db;

import java.time.Instant;
import java.util.Collections;
import java.util.Date;
import java.util.Set;
import java.util.TreeSet;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.persistence.PreRemove;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;

import org.hibernate.annotations.SortNatural;

import com.bnorm.barkeep.model.Book;
import com.bnorm.barkeep.model.Recipe;

@Entity
@Table(name = "tblBooks")
@NamedQueries({@NamedQuery(name = "BookEntity.findAll", query = "SELECT b FROM BookEntity b")})
public class BookEntity implements Book {

  @Id
  @GeneratedValue
  @Column(name = "id", unique = true, nullable = false)
  private long id;

  @ManyToOne
  @JoinColumn(name = "owner", referencedColumnName = "id", nullable = false)
  private UserEntity owner;

  @Column(name = "title", nullable = false)
  private String title;

  @Column(name = "description")
  private String description;

  @Column(name = "active", nullable = false)
  private boolean active;

  @Temporal(TemporalType.TIMESTAMP)
  @Column(name = "createTime", updatable = false)
  private Date createTime;

  @Temporal(TemporalType.TIMESTAMP)
  @Column(name = "modifyTime", updatable = false)
  private Date modifyTime;

  @ManyToMany(mappedBy = "books")
  @SortNatural
  private Set<UserEntity> users = new TreeSet<>();

  @ManyToMany
  @JoinTable(name = "lkpBooksRecipes",
             joinColumns = @JoinColumn(name = "book"),
             inverseJoinColumns = @JoinColumn(name = "recipe"))
  @SortNatural
  private Set<RecipeEntity> recipes = new TreeSet<>();

  @PreRemove
  public void onRemove() {
    for (UserEntity userEntity : users) {
      userEntity.removeBook(this);
    }
  }

  public BookEntity() {
  }

  @Override
  public Long getId() {
    return id;
  }

  @Override
  public UserEntity getOwner() {
    return owner;
  }

  public void setOwner(UserEntity owner) {
    this.owner = owner;
  }

  @Override
  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  @Override
  public String getDescription() {
    return description;
  }

  public boolean isActive() {
    return active;
  }

  public void setActive(boolean active) {
    this.active = active;
  }

  public void setDescription(String description) {
    this.description = description;
  }

  public Instant getCreateTime() {
    return createTime.toInstant();
  }

  public Instant getModifyTime() {
    return modifyTime.toInstant();
  }

  @Override
  public Set<Recipe> getRecipes() {
    return Collections.unmodifiableSet(recipes);
  }

  public void addRecipe(RecipeEntity recipe) {
    recipes.add(recipe);
  }
}

File: model-database/src/main/kotlin/com/bnorm/barkeep/db/BookEntity.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db

import com.bnorm.barkeep.model.Book
import org.hibernate.annotations.SortNatural
import java.time.LocalDateTime
import java.util.TreeSet
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.NamedQueries
import javax.persistence.NamedQuery
import javax.persistence.PreRemove
import javax.persistence.Table

@Entity
@Table(name = "tblBooks")
@NamedQueries(NamedQuery(name = "BookEntity.findAll", query = "SELECT b FROM BookEntity b"))
class BookEntity : Book {

  @Id
  @GeneratedValue
  @Column(name = "id", unique = true, nullable = false)
  override var id: Long = 0

  @ManyToOne
  @JoinColumn(name = "owner", referencedColumnName = "id", nullable = false)
  override var owner: UserEntity? = null

  @Column(name = "title", nullable = false)
  override var title: String? = null

  @Column(name = "description")
  override var description: String? = null

  @Column(name = "active", nullable = false)
  var isActive: Boolean = false

  @Column(name = "createTime", updatable = false)
  private val createTime: LocalDateTime? = null

  @Column(name = "modifyTime", updatable = false)
  private val modifyTime: LocalDateTime? = null

  @ManyToMany(mappedBy = "books")
  @SortNatural
  private val users: MutableSet<UserEntity> = TreeSet()

  @ManyToMany
  @JoinTable(name = "lkpBooksRecipes",
             joinColumns = arrayOf(JoinColumn(name = "book")),
             inverseJoinColumns = arrayOf(JoinColumn(name = "recipe")))
  @SortNatural
  override val recipes: MutableSet<RecipeEntity> = TreeSet()

  @PreRemove
  fun onRemove() {
    for (userEntity in users) {
      userEntity.removeBook(this)
    }
  }

  fun addRecipe(recipe: RecipeEntity) {
    recipes.add(recipe)
  }
}
